File: ruby/request.rb
require 'dotenv'
require 'json'
require 'jwt'
require 'rest-client'
require 'base64'

Dotenv.load('../.env')

module BaseRequestRoute

  DEBUG       = ENV['DEBUG']&.downcase == 'true'
  ACCESS_KEY  = ENV['ACCESS_KEY']
  SECRET_KEY  = ENV['SECRET_KEY']
  BASE_URL    = ENV['API_URL']

  def self.request(method, path, query_params: {}, body: '')
    p "#{method.to_s.upcase} #{path} query_params=#{query_params} body=#{body}" if DEBUG

    payload = {
      path: path,
      content: Digest::SHA2.hexdigest(body)
    }

    url = "#{BASE_URL}#{path}"

    headers = { access_key: ACCESS_KEY }

    token = JWT.encode(
      payload,
      SECRET_KEY,
      'HS256',
      headers
    )

    p "token=#{token}" if DEBUG

    if method == :get && !query_params.empty?
      query_string = []

      query_params.each do |key, val|
        case val
        when Array
          val.each do |v|
            query_string << "#{key}[]=#{URI.encode_www_form_component(v)}"
          end
        when Hash
          serialized_hash = URI.encode_www_form_component(val.to_json)
          query_string << "#{key}=#{serialized_hash}"
        else
          query_string << "#{key}=#{URI.encode_www_form_component(val)}"
        end
      end
      url = "#{BASE_URL}#{path}?#{query_string.join('&')}"
    end

    begin
        res = RestClient::Request.execute(
          method: method,
          url: url,
          payload: body,
          headers: {
            authorization: "Bearer #{token}",
            content_type: :json
          }
        )

        puts "Response status code: #{res.code}"

        request_name = method.to_s + path.gsub('/', '_')

        if res.headers[:content_type] == 'image/png'
          File.open("./documents/#{request_name}_response.png", 'wb') do |file|
            file.write(res.body)
          end
          puts "Image saved as #{request_name}_response.png"
        elsif res.headers[:content_type] == 'application/pdf'
          File.open("./documents/#{request_name}_response.pdf", 'wb') do |file|
            file.write(res.body)
          end
          puts "PDF saved as get_#{request_name}_response.pdf"
    else
      puts JSON.pretty_generate(JSON.parse(res.body))
      JSON.parse(res.body)
    end
    rescue StandardError => error
      puts error
      puts error.http_body if defined?(error.http_body) && error.to_s != '500 Internal Server Error'
      puts error.res if defined?(error.res)
      end
    end
end

File: python/request.py
import logging
import hashlib
import json
from multiprocessing.sharedctypes import Value
import jwt
import requests
import urllib.parse
from dotenv import dotenv_values

config = dotenv_values()

class BaseRequest:
  def __init__(self):
    self.access_key = config['ACCESS_KEY']
    self.secret_key = config['SECRET_KEY']
    self.api_url = config['API_URL']
    self.logger = logging.getLogger(__name__)

  def generate_token(self, path, payload):
    headers = {"access_key": self.access_key}
    payload_data = {
      "path": path,
      "content": hashlib.sha256(payload.encode('ascii')).hexdigest()
    }
    return jwt.encode(payload_data, self.secret_key, algorithm='HS256', headers=headers)
  
  def make_request(self, method, path, params=None, body=None, headers=None):
    if headers is None:
      headers = {}
    if params and body:
      raise ValueError("Both 'params' and 'body' cannot be provided simultaneously. Please modify request.")

    url = self.api_url + path
    payload = ""

    if method == "GET" and params:
        query_string = []
        for key, value in params.items():
            if isinstance(value, list):
                for v in value:
                    query_string.append(f"{key}[]={urllib.parse.quote(v)}")
            elif isinstance(value, dict):
                serialized_value = urllib.parse.quote(json.dumps(value))
                query_string.append(f"{key}={serialized_value}")
            else:
                query_string.append(f"{key}={urllib.parse.quote(str(value))}")
        query_params = '&'.join(query_string)
        url = f"{url}?{query_params}"
        response = requests.get(url, headers=headers)
    else:
        if body:
          payload = json.dumps(body)
    
    try:
      url = url
      headers["Authorization"] = 'Bearer %s' % self.generate_token(path, payload)
      headers["Content-Type"] = "application/json"
      print(f'{method} {path} body={body}')
      response = requests.request(method, url, headers=headers, data=payload)
      request_name = method.lower() + path.replace('/', '_')

      if response.headers.get("content-type") == "application/json":
        return response.json()
      elif response.headers.get("content-type") == "image/png":
        with open(f'./documents/{request_name}_response.png', 'wb') as file:
            file.write(response.content)
        print(f'PNG image saved as {request_name}_response.png')
        return "PNG image downloaded."
      elif response.headers.get("content-type") == "application/pdf":
        with open(f'./documents/{request_name}_response.pdf', 'wb') as file:
            file.write(response.content)
        print(f'PDF file saved as {request_name}_response.pdf')
        return "PDF file downloaded"
      else:
        return response.text

    except Exception as e:
      self.logger.error(msg="{e}".format(e=e))
      raise e
